### PR TITLE
Fix rich with uneven refresh rate tracking

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

2 fixes here:
- One when refresh_rate isn't aligned with progress total, for eg. if total is 11 and refresh_rate is 5, it doesn't end up with the correct total.
- When using val_check_interval, rich validation bar shows total as the total number of validation steps going to happen during a single epoch, rather it should only show the total for one validation loop just like tqdm

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @SeanNaren @kaushikb11

### CODE REVIEW

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -346,32 +344,35 @@ def on_train_epoch_start(self, trainer, pl_module):
...
         if self.progress is not None:
             return self.progress.add_task(
                 f"[{self.theme.description}]{description}", total=total_batches, visible=visible
             )
 
-    def _update(self, progress_bar_id: int, current: int, total: int, visible: bool = True) -> None:
+    def _update(self, progress_bar_id: int, current: int, total: Union[int, float], visible: bool = True) -> None:
         if self.progress is not None and self._should_update(current, total):
-            self.progress.update(progress_bar_id, advance=self.refresh_rate, visible=visible)
+            left_over = current % self.refresh_rate
```
Comment (carmocca) [inline pytorch_lightning/callbacks/progress/rich_progress.py:364]: nit:

```suggestion
            leftover = current % self.refresh_rate
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 2]
```diff
@@ -178,9 +178,6 @@ def _get_max_batches(self) -> List[int]:
             max_batches = self.trainer.num_test_batches
         else:
             if self.trainer.sanity_checking:
-                self.trainer.num_sanity_val_batches = [
-                    min(self.trainer.num_sanity_val_steps, val_batches) for val_batches in self.trainer.num_val_batches
-                ]
```
Comment (carmocca) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:183]: Why is this moved?

File: tests/callbacks/test_rich_progress_bar.py [hunk 3]
```diff
@@ -61,25 +60,41 @@ def test_dataloader(self):
...
 
+    def _initialize_trainer():
+        trainer = Trainer(
+            default_root_dir=tmpdir,
+            num_sanity_val_steps=0,
+            limit_train_batches=1,
+            limit_val_batches=1,
+            limit_test_batches=1,
+            limit_predict_batches=1,
+            max_steps=1,
```
Comment (carmocca) [inline tests/callbacks/test_rich_progress_bar.py:71]: Not required

```suggestion
```

**Review by carmocca — APPROVED** [review]
Looks good. Do we need this in a bug-fix release?

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:183]: no reason, just for better readability. since setting it inside `_get_max_batches` doesn't makes sense.

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 4]
```diff
@@ -346,32 +344,35 @@ def on_train_epoch_start(self, trainer, pl_module):
...
+            advance = left_over if (current == total and left_over != 0) else self.refresh_rate
```
Comment (rohitgr7) [inline pytorch_lightning/callbacks/progress/rich_progress.py:365]:
```suggestion
            leftover = current % self.refresh_rate
            advance = left_over if (current == total and leftover != 0) else self.refresh_rate
```

**Comment by rohitgr7** [comment]
> Looks good. Do we need this in a bug-fix release?

yes, if we are going to make one.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:183]: Sounds good to me if it works. I remember I tried moving this a long time ago but the structure did not allow it. Maybe it's different now

**Review by SeanNaren — APPROVED** [review]
Thanks @rohitgr7!

**Review by ethanwharris — APPROVED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -457,6 +457,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Disbled sampler replacement when using `IterableDataset` ([#11507](https://github.com/PyTorchLightning/pytorch-lightning/pull/11507))
 
 
+- Fixed `RichProgressBar` progress with uneven refresh rate ([#11507](https://github.com/PyTorchLightning/pytorch-lightning/pull/11668))
+
+
+- Fixed `RichProgressBar` progress validation bar total when using multiple validation runs within a single training epoch ([#11507](https://github.com/PyTorchLightning/pytorch-lightning/pull/11668))
```
Comment (akihironitta) [inline CHANGELOG.md:463]:
```suggestion
- Fixed `RichProgressBar` progress with uneven refresh rate ([#11668](https://github.com/PyTorchLightning/pytorch-lightning/pull/11668))


- Fixed `RichProgressBar` progress validation bar total when using multiple validation runs within a single training epoch ([#11668](https://github.com/PyTorchLightning/pytorch-lightning/pull/11668))
```

Comment (rohitgr7) [reply CHANGELOG.md:463]: good eye 🦅

File: pytorch_lightning/callbacks/progress/rich_progress.py [hunk 6]
```diff
@@ -346,32 +344,35 @@ def on_train_epoch_start(self, trainer, pl_module):
...
 
     def _add_task(self, total_batches: int, description: str, visible: bool = True) -> Optional[int]:
         if self.progress is not None:
             return self.progress.add_task(
                 f"[{self.theme.description}]{description}", total=total_batches, visible=visible
             )
 
-    def _update(self, progress_bar_id: int, current: int, total: int, visible: bool = True) -> None:
+    def _update(self, progress_bar_id: int, current: int, total: Union[int, float], visible: bool = True) -> None:
         if self.progress is not None and self._should_update(current, total):
```
Comment (Borda) [inline pytorch_lightning/callbacks/progress/rich_progress.py:363]: maybe reverse logic, also easier if you need to put breakpoint to return...
```py
        if self.progress is None or not self._should_update(current, total):
            return
        leftover = current % self.refresh_rate
        advance = leftover if (current == total and leftover != 0) else self.refresh_rate
        self.progress.update(progress_bar_id, advance=advance, visible=visible)
        self.refresh()
```

Comment (rohitgr7) [reply pytorch_lightning/callbacks/progress/rich_progress.py:363]: I always wonder, is this a better code pattern from what's already there?

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -506,10 +506,16 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue in `RichProgressbar` to display the metrics logged only on main progress bar ([#11690](https://github.com/PyTorchLightning/pytorch-lightning/pull/11690))
 
 
-- Fixed check for available modules ([#11526](https://github.com/PyTorchLightning/pytorch-lightning/pull/11526))
+- Fixed `RichProgressBar` progress with uneven refresh rate ([#11668](https://github.com/PyTorchLightning/pytorch-lightning/pull/11668))
```
Comment (rohitgr7) [inline CHANGELOG.md:509]:
```suggestion
- Fixed `RichProgressBar` progress when refresh rate does not evenly divide the total counter ([#11668](https://github.com/PyTorchLightning/pytorch-lightning/pull/11668))
```
